### DISCUSSION TITLE
Calculation of permeability

### DISCUSSION THREAD

**Discussion by patiln8** [PMEAL/OpenPNM#2827] [Q&A]
We have a porous data with fibers and binders of size 500x500x500 and would like to do REV analysis and find out the permeability. So how can this be done? 
I am using this code as a reference : https://github.com/mbandreeta/absolute_permeability_simulations/blob/main/permeability_PoreSpy_OpenPNM.ipynb

When I try to create network and add the model. I get an error as follows: 
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File /opt/conda/lib/python3.9/site-packages/openpnm/core/_base2.py:173, in Base2.__getitem__(self, key)
    172 try:
--> 173     return super().__getitem__(key)
    174 except KeyError:
    175     # If key is object's name or all, return ones

KeyError: 'throat.hydraulic_size_factors'

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
Input In [64], in <module>
     15 water = op.phase.Water(network=pn)
     17 #conductance -- ????
---> 18 water.add_model(propname='throat.hydraulic_conductance', model=op.models.physics.hydraulic_conductance.hagen_poiseuille)
     20 #Stokes
     21 perm = op.algorithms.StokesFlow(network=pn, phase=water)

File /opt/conda/lib/python3.9/site-packages/openpnm/core/_models.py:404, in ModelsMixin2.add_model(self, propname, model, domain, regen_mode, **kwargs)
    402 self.models[propname+'@'+domain] = ModelWrapper(**kwargs)
    403 if regen_mode != 'deferred':
--> 404     self.run_model(propname+'@'+domain)

File /opt/conda/lib/python3.9/site-packages/openpnm/core/_models.py:502, in ModelsMixin2.run_model(self, propname, domain)
    500 if '@' in propname:  # Get domain from propname if present
    501     propname, _, domain = propname.partition('@')
--> 502     self.run_model(propname=propname, domain=domain)
    503 else:  # No domain means run model for ALL domains
    504     for item in self.models.keys():

File /opt/conda/lib/python3.9/site-packages/openpnm/core/_models.py:521, in ModelsMixin2.run_model(self, propname, domain)
    519 if 'domain' not in inspect.getfullargspec(mod_dict['model']).args:
    520     _ = kwargs.pop('domain', None)
--> 521     vals = mod_dict['model'](self, **kwargs)
    522     if isinstance(vals, dict):  # Handle models that return a dict
    523         for k, v in vals.items():

File /opt/conda/lib/python3.9/site-packages/openpnm/models/physics/hydraulic_conductance/_funcs.py:81, in hagen_poiseuille(phase, pore_viscosity, throat_viscosity, size_factors)
     56 @_doctxt
     57 def hagen_poiseuille(
     58     phase,
   (...)
     61     size_factors="throat.hydraulic_size_factors"
     62 ):
     63     r"""
     64     Calculates the hydraulic conductance of conduits in network.
     65 
   (...)
     79 
     80     """
---> 81     return generic_hydraulic(phase=phase,
     82                              pore_viscosity=pore_viscosity,
     83                              throat_viscosity=throat_viscosity,
     84                              size_factors=size_factors)

File /opt/conda/lib/python3.9/site-packages/openpnm/models/physics/hydraulic_conductance/_funcs.py:42, in generic_hydraulic(phase, pore_viscosity, throat_viscosity, size_factors)
     39 mu1, mu2 = phase[pore_viscosity][conns].T
     40 mut = phase[throat_viscosity]
---> 42 SF = network[size_factors]
     43 if isinstance(SF, dict):  # Legacy approach
     44     F1, Ft, F2 = SF.values()

File /opt/conda/lib/python3.9/site-packages/openpnm/core/_base2.py:188, in Base2.__getitem__(self, key)
    186     return vals
    187 else:
--> 188     raise KeyError(key)

KeyError: 'throat.hydraulic_size_factors' 

How can this be reolved?

**Comment by jgostick** [comment]
The error `KeyError: 'throat.hydraulic_size_factors'` means that you have not defined a pore-scale model that computes the hydraulic_size_factor.  You can add this yourself with something like `pn.add_model(propname='throat.hydraulic_size_factor', model=op.models.geometry.hydaulic_size_factors.spheres_and_cylinders)` or what ever model suits your assumed geometry.

**Comment by patiln8** [comment]
I have defined the model as: 
water.add_model(propname='throat.hydraulic_size_factors', model=op.models.geometry.hydraulic_size_factors.spheres_and_cylinders(network=pn, pore_diameter= pn.Ps, throat_diameter=pn.Ts))

But I get the following error:
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [22], in <module>
     15 water = op.phase.Water(network=pn)
     17 #conductance -- ????
---> 18 water.add_model(propname='throat.hydraulic_size_factors', model=op.models.geometry.hydraulic_size_factors.spheres_and_cylinders(network=pn, pore_diameter= pn.Ps, throat_diameter=pn.Ts))
     20 #Stokes
     21 perm = op.algorithms.StokesFlow(network=pn, phase=water)

File /opt/conda/lib/python3.9/site-packages/openpnm/models/geometry/hydraulic_size_factors/_funcs.py:62, in spheres_and_cylinders(network, pore_diameter, throat_diameter)
     24 @_geodocs
     25 def spheres_and_cylinders(
     26     network,
     27     pore_diameter="pore.diameter",
     28     throat_diameter="throat.diameter",
     29 ):
     30     r"""
     31     Computes hydraulic size factors for conduits assuming pores are
     32     spheres and throats are cylinders.
   (...)
     60 
     61     """
---> 62     D1, Dt, D2 = network.get_conduit_data(pore_diameter.split('.', 1)[-1]).T
     63     L1, Lt, L2 = _conduit_lengths.spheres_and_cylinders(
     64         network=network,
     65         pore_diameter=pore_diameter,
     66         throat_diameter=throat_diameter
     67     ).T
     69     # Fi is the integral of (1/A^2) dx, x = [0, Li]

AttributeError: 'numpy.ndarray' object has no attribute 'split'

**Comment by jgostick** [comment]
You've "run" the model instead of passing in only a handle to the model.  The difference is:

This: 

```python
pn.add_model(propname='throat.hydraulic_size_factor',
             model=op.models.geometry.hydaulic_size_factors.spheres_and_cylinders) 
```

Vs:

```python
water.add_model(propname='throat.hydraulic_size_factors', 
                model=op.models.geometry.hydraulic_size_factors.spheres_and_cylinders(
                    network=pn, 
                    pore_diameter=pn.Ps, 
                    throat_diameter=pn.Ts))
```

You're also adding to the water object instead of the network object.  I'm sorry that OpenPNM is a bit complex to use sometimes, but please pay a bit more attention to each line before giving up.

**Comment by andtoni** [comment]
Hi Prof Gostick

I am also getting the KeyError: 'throat.hydraulic_size_factors'

Apologies for opening this up again but I am trying to follow exactly as you suggest here, but still get a "diameter not found" key error with this code:
pn.add_model(propname = 'throat.hydraulic_size_factors',
             model=op.models.geometry.hydraulic_size_factors.spheres_and_cylinders)
